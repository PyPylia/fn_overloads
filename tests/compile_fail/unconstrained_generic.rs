#![feature(unboxed_closures, fn_traits)]
use fn_overloads::fn_overloads;
use std::{ops::Mul, str::FromStr};

fn_overloads! {
    fn double {
        <R: FromStr + Mul<u32>>(value: &str) -> Option<R> {
            Some(value.parse::<R>().ok()? * 2)
        }
    }
}

struct CustomMul;
impl Mul<&'static str> for CustomMul {
    type Output = String;

    fn mul(self, rhs: &'static str) -> Self::Output {
        rhs.to_string()
    }
}

fn main() {
    double("asdf");
}