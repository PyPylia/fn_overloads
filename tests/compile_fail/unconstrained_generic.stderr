error[E0207]: the type parameter `R` is not constrained by the impl trait, self type, or predicates
 --> tests/compile_fail/unconstrained_generic.rs:7:10
  |
7 |         <R: FromStr + Mul<u32>>(value: &str) -> Option<R> {
  |          ^ unconstrained type parameter
